src/core/scope.ml

```diff
@@ -135,22 +132,22 @@ let get_root : p_term -> sig_state -> sym * pp_hint = fun t ss ->
 (** Representation of the different scoping modes.  Note that the constructors
     hold specific information for the given mode. *)
 type mode =
-  | M_Term of metamap Stdlib.ref * expo
+  | M_Term of meta StrMap.t Stdlib.ref * expo
   (** Standard scoping mode for terms, holding a map of metavariables that can
       be updated with new metavariables on scoping and the exposition of the
       scoped term. *)
   | M_Patt
   (** Scoping mode for patterns in the rewrite tactic. *)
-  | M_LHS  of (string * int) list * bool
+  | M_LHS  of (string * int) list * bool * int Stdlib.ref
```

Comment:
I don't understand the role of the new reference here. Can you explain? What problem does that solve?
In particular, I don't see what implicit arguments have to do with anything here.

Reply:
I can only review the rest of the file once I understand what you want to do here.

Reply:
Now, you can also have implicit arguments in LHS too! :-)
For instance, with polymorphic lists, you can simply write concat (cons x l) m -> cons x (concat l m)!

Reply:
Wasn't it already the case?? For instance, I can pick the following rewriting rule
```
symbol lif {U: Set} {V: Set} (S: Set)
: ε (U ⊑ S) ⇒ ε (V ⊑ S) ⇒ Bool ⇒ Deferred.t U ⇒ Deferred.t V ⇒ η S

// lazy if
rule lif _ &pr _ true &t _ → ↑ &pr (Deferred.unquote &t)
```

Reply:
In your example, the RHS does not depend on the implicit variables of the LHS. In general, implicit variables of the LHS needs to be used in the RHS as well.

Reply:
I think it does, if I call with `--verbose 3` and flag implicits on, I obtain
```
lif &#0 &#1 &#2 &pr &#3 true &t &#4 → ↑ ?#0 ?#2 &pr (Deferred.unquote ?#0 &t)
```
the upcast ↑ has implicits as well. Here, the first implicit of the upcast is the first implicit of the `lif`

Reply:
Yeah, implicits were already handled correctly I think. I agree with @gabrielhdt.

Reply:
No! It was completely buggy. As you can see in your example: the RHS of the rule added in the signature contains uninstantiated metavariables (identifiers prefixed by '?')!!! The master branch doesn't complain but it is wrong. ?#0 should be replaced by say &#0, etc.

Reply:
So in your version, we have
```
lif &#0 &#1 &#2 &pr &#3 true &t &#4 → ↑ &#0 &#2 &pr (Deferred.unquote &#0 &t)
```
?

Reply:
But why should it be any different to meta-variables in the domains of lambdas? I think that it should be the role of SR checking to instantiate such meta variables with the proper thing. For now what do you generate exactly instead of meta variables?

Reply:
Morally, this could be any term depending on the higher-order variables bound in the LHS.

Reply:
@gabrielhdt if you give me a standalone file, I can tell you. You can also check the output by yourself on my branch.

Reply:
Here is an example;
```
set flag "print_implicits" on
set verbose 3

// Type of set codes

constant symbol Set : TYPE

// Interpretation of set codes in TYPE

set declared "τ"

injective symbol τ : Set ⇒ TYPE

// Type of natural numbers

constant symbol nat : Set

definition N ≔ τ nat

constant symbol zero : N
constant symbol succ : N ⇒ N

// Type of polymorphic lists

constant symbol list : Set ⇒ Set

definition L a ≔ τ (list a)

set declared "□"

constant symbol □ {a} : L a
constant symbol cons {a} : τa ⇒ L a ⇒ L a

set infix 4 "⸬" ≔ cons

// Length of a list

symbol length {a} : L a ⇒ N

rule length □ → zero
and length (_ ⸬ &l) → succ (length &l)

// Concatenation of two lists

symbol concat {a} : L a ⇒ L a ⇒ L a

set infix right 5 "⋅" ≔ concat

rule □ ⋅ &m → &m
and (&x ⸬ &l) ⋅ &m → &x ⸬ (&l ⋅ &m)
```
master branch output:
```
(flag) verbose → 3     
(symb) Set
(conf) declared identifier [τ]
(symb) τ
(symb) nat
(symb) N ≔ τ nat
(symb) zero
(symb) succ
(symb) list
(symb) L ≔ λa, τ (list a)
(conf) declared identifier [□]
(symb) □
(symb) cons
(conf) new infix [⸬]
(symb) length
(rule) length &#0 (□ &#1) → zero
(rule) length &#0 ((⸬) &#1 &#2 &l) → succ (length ?#0 &l)
(symb) concat
(conf) new infix [⋅]
(rule) (⋅) &#0 (□ &#1) &m → &m
(rule) (⋅) &#0 ((⸬) &#1 &x &l) &m → (⸬) ?#0 &x ((⋅) ?#0 &l &m)
Checked [/home/blanqui/src/lambdapi/tests/OK/328.lp]
```
rhs_metas branch output:
```
(flag) verbose → 3     
(symb) Set
(conf) declared identifier [τ]
(symb) τ
(symb) nat
(symb) N ≔ τ nat
(symb) zero
(symb) succ
(symb) list
(symb) L ≔ λa, τ (list a)
(conf) declared identifier [□]
(symb) □
(symb) cons
(conf) new infix [⸬]
(symb) length
(rule) length &#0 (□ &#1) → zero
(rule) length &#1 ((⸬) &#2 &#3 &l) → succ (length &#1 &l)
(symb) concat
(conf) new infix [⋅]
(rule) (⋅) &#1 (□ &#2) &m → &m
(rule) (⋅) &#3 ((⸬) &#4 &x &l) &m → (⸬) &#3 &x ((⋅) &#3 &l &m)
```

Reply:
All right, my example is getting more and more involved (I'm working on it), and actually it requires your branch (the master raises `assert false`).

Reply:
@fblanqui I really do not get it, the LHS is fine on the master branch. The problem is with the RHS. However one thing that seems like a regression with your version is that the numbering of pattern variables seems somewhat broken. There is no `#0` or `#1` in the last rule for example.

The existing problem had to do with the RHS, not the LHS. And it indeed seems to have been solved by your PR. However the changes for the LHS are most certainly not what we want. Am I wrong there?

Reply:
They are just names. It should not be too difficult to fix. This is not a reason to reject this PR.

Reply:
OK so I did some tests, and I am coming back to something I said long ago when we started to work on implicit arguments: I think that the LHS of rules should always have their implicit arguments explicitly given.

I am starting to understand what you tried to do here @fblanqui. Any pattern variable of the LHS (including those that stand for a wildcard) can actually appear in the RHS through unification of meta-variables corresponding to implicit arguments or wildcards in the RHS. So the optimisation that consisted in only binding the pattern variables that at least appear explicitly in the RHS or are non-left-linear is wrong. We should really bind all the variables, or at least wait until after SR-checking to do the optimisation rightly.

Reply:
The reason why it starts at 3 is because the slots 0 to 2 are taken by &x, &l and &m.

Reply:
Exactly. But I don't think that this optim is really useful.

Reply:
> They are just names. It should not be too difficult to fix. This is not a reason to reject this PR.

Well, nobody is talking about rejecting the PR. This is a regression, it should probably be taken care of while we are at it.

> The reason why it starts at 3 is because the slots 0 to 2 are taken by &x, &l and &m.

I see. It is still weird that the generation of numbers does not go from left to right in the LHS. In any case, I would change printing to get "&%i" when no name, and "&%i_%s" when there is a name, that would be clearer and useful for debugging.

> Exactly. But I don't think that this optim is really useful.

OK, I kind of agree with that. In that case, you should turn the `int option` argument of the `Patt` constructor into an `int`. And the code of rule scoping can probably be simplified a lot. I can help with that if you want.

Reply:
> I have a question: do we assume somewhere in he code that a meta of arity n has a type of the form \forall x1:A1,..,\forall xn:An,B?

I'm not sure, but most probably.

Reply:
It would affect tree.ml. In particular, None is used in tree.ml line 529. I propose that you or Gabriel (this is his code) do this modification in another PR.

Reply:
I am looking into it at the moment actually.

Reply:
As you want but I prefer to merge the PR as it is now for not mixing these modifications with mine. Modifying tree.ml will require more extensive testing. And this PR is waited for by various people. So, if you have no other comment on the current code, then I will merge.

Reply:
So here are my conclusions after working late on the problem:
 - Removing the option on `Patt` naively does not work, and in fact we should better keep it.
 - I want to keep it so that we can optimise rules after SR-checking. At that point we know what are the pattern variables that are useful and those that are not, so we can apply a transformation there. What we did before was incorrect and would definitely lead to bugs.
 - What I did is define a notion of pre-rule, that contains some extra useful information. SR-checking takes a pre-rule and returns a rule. This way we have all the freedom to optimise or not. And we can also get some useful informations that simplify the implementation of SR-checking.
 - I have basically finished the work, but I want to make sure that I now use Bindlib correctly. The current PR as it stands has quite a few bugs that I exposed during refactoring (missing cases, wrong bindings). I now know how to solve the problem completely, and I will finish my work tomorrow.
 - Roughly, the problem is again that the types of metavariables in a RHS can contain free term environment variables, and lifting assumes that metavariables are closed (including their type). The solution is to do more lifting in some cases.

Reply:
Here is what I have done so far: https://github.com/rlepigre/lambdapi/commit/d112faaef8e9602235c866eb1859b8a79f6816f8.

Reply:
I am done @fblanqui. If you agree to the changes in https://github.com/rlepigre/lambdapi/commit/790c0d294735df4ecd4ffb7adca386c71948896d then I will push them here and merge your PR. The performances are almost identical with my changes (although I observed a small, not very significant speedup on both `make tests` and `make focalide`). I will implement the optimisation of rules in another PR, for now there is only a `TODO` in `src/core/sr.ml`.

Reply:
Note that there is quite a bit more code, but this is necessary to handle variable binding correctly and with less chance of having bugs. The thing you did updating the types of metavariables was certainly not correct in general. Now we have more general scoping for rules, and sanity checks to ensure that metavariables outside of the checking of a RHS cannot contain variables other than those of the environment.

Reply:
Thanks to check your commit on https://github.com/fblanqui/lib/.

Reply:
Thanks to merge the last commits I did on rhs_metas too.

Reply:
All merged, and the library still works fine.